### PR TITLE
ARROW-3442: [C++] Allow dynamic linking of (most) unit tests

### PR DESCRIPTION
**Pull request by pitrou**
Use dynamic linking in preference for tests, as it reduces the disk footprint and probably makes linking faster too.

### CODE REVIEW

File: cpp/src/parquet/statistics-test.cc [hunk 1]
```diff
@@ -381,26 +385,16 @@ TEST(CorrectStatistics, Basics) {
...
+  parquet::WriterProperties::Builder builder;
+  builder.created_by(created_by);
+  std::shared_ptr<parquet::WriterProperties> props = builder.build();
+
+  AssertStatsSet(version, props, schema.Column(0), true);
+  AssertStatsSet(version, props, schema.Column(1), true);
+  AssertStatsSet(version, props, schema.Column(2), true);
+  AssertStatsSet(version, props, schema.Column(3), true);
+  AssertStatsSet(version, props, schema.Column(4), false);
+  AssertStatsSet(version, props, schema.Column(5), true);
```
Comment (wesm) [inline cpp/src/parquet/statistics-test.cc]: Nice simplification

**Review by wesm — APPROVED** [review]
+1, thank you for doing this! I'll take this for a spin locally also on Ubuntu 14.04 with both gcc 4.8.5 and clang 6

**Comment by pitrou** [comment]
There is a problem, though: the Plasma Java client tests are broken:
https://travis-ci.org/apache/arrow/jobs/440089184#L3463

This is because `plasma_store_server` is now dynamically linked, and for some reason an rpath doesn't seem set for `libplasma.so.12` and `libarrow.so.12`. I don't understand how that happens, as it does have an rpath when I build it here.

We could work around this by building `plasma_store_server`, but it would be nice to understand what happens. Unfortunately I don't know much about the rpath mechanisms. @xhochy

**Comment by pitrou** [comment]
Ah, it seems we set LD_LIBRARY_PATH at the beginning of `travis_script_python.sh`. We could do the same at the beginning of `travis_script_plasma_java_client.sh`. Does that sound ok?

**Comment by wesm** [comment]
Yes, that's OK!

**Comment by wesm** [comment]
Since we have static libs in the toolchain I re-enabled re2 here. Everything looks good for me except that the build doesn't work with gcc 4.8.5 on account of Gandiva libraries using `-static-libstdc++ -static-libgcc` for which there is already a JIRA to fix

**Comment by xhochy** [comment]
That's ok for now. I would prefer not have `LD_LIBRARY` in the tests but I can look into that some time. We should though never require `LD_LIBRARY_PATH` in installed executables.

**Comment by pitrou** [comment]
AppVeyor build at https://ci.appveyor.com/project/pitrou/arrow/builds/19438131

**Pull request closed without merging**